File: guess.cpp
#include<cstdio>
#include<cstdlib>
#include<fstream>
#include<iostream>
#include<map>
#include<set>
#include<string>
#include<vector>

using namespace std;

int MONEY_LARGE = 0, MONEY_BIG = 2250, MONEY_MEDIUM = 1000, MONEY_SMALL = 450;
int CREDIT_BIG = 105, CREDIT_SMALL = 70;

int dif_to_points(int dif)
{
	return (dif * dif - 15 * dif + 58) / 2;
}

int legend_points(int points, int trainer_number)
{
	if (trainer_number <= 8)
		return (points >= 203);
	if (trainer_number == 9 || trainer_number == 10)
		return (points >= 200);
	if (trainer_number == 11 || trainer_number == 12)
		return (points >= 190);
	if (trainer_number >= 13)
		return (points >= 180);
}

int pokemon_points(int points, int trainer_number)
{
	if (trainer_number <= 8)
		return (points >= 190);
	if (trainer_number == 9 || trainer_number == 10)
		return (points >= 180);
	if (trainer_number == 11 || trainer_number == 12)
		return (points >= 165);
	if (trainer_number >= 13)
		return (points >= 150);
}

int points_to_money(int points, int trainer_number)
{
	if (trainer_number <= 8)
	{
		if (points >= 203)
			return 0;
		if (points >= 190)
			return MONEY_LARGE;
		if (points >= 175)
			return MONEY_BIG;
		if (points >= 165)
			return MONEY_MEDIUM;
		if (points >= 155)
			return MONEY_SMALL;
	}
	if (trainer_number == 9 || trainer_number == 10)
	{
		if (points >= 200)
			return 0;
		if (points >= 180)
			return MONEY_LARGE;
		if (points >= 160)
			return MONEY_BIG;
		if (points >= 150)
			return MONEY_MEDIUM;
		if (points >= 140)
			return MONEY_SMALL;
	}
	if (trainer_number == 11 || trainer_number == 12)
	{
		if (points >= 190)
			return 0;
		if (points >= 165)
			return MONEY_LARGE;
		if (points >= 140)
			return MONEY_BIG;
		if (points >= 130)
			return MONEY_MEDIUM;
		if (points >= 120)
			return MONEY_SMALL;
	}
	if (trainer_number >= 13)
	{
		if (points >= 180)
			return 0;
		if (points >= 150)
			return MONEY_LARGE;
		if (points >= 120)
			return MONEY_BIG;
		if (points >= 110)
			return MONEY_MEDIUM;
		if (points >= 100)
			return MONEY_SMALL;
	}
	return 0;
}

int points_to_credit(int points, int trainer_number)
{
	if (trainer_number <= 8)
	{
		if (points >= 190)
			return CREDIT_BIG;
		if (points >= 175)
			return CREDIT_SMALL;
	}
	if (trainer_number == 9 || trainer_number == 10)
	{
		if (points >= 180)
			return CREDIT_BIG;
		if (points >= 160)
			return CREDIT_SMALL;
	}
	if (trainer_number == 11 || trainer_number == 12)
	{
		if (points >= 165)
			return CREDIT_BIG;
		if (points >= 140)
			return CREDIT_SMALL;
	}
	if (trainer_number >= 13)
	{
		if (points >= 150)
			return CREDIT_BIG;
		if (points >= 120)
			return CREDIT_SMALL;
	}
	return 0;
}


int main()
{	
	ifstream cwin("winners.txt", std::ifstream::in);
	vector<int> trainer_number(2);
	vector<string> empty(8);
	vector<vector<string> > winner(2, empty);
	for (int i = 0; i < 2; ++i)
	{
		cwin>>trainer_number[i];
		for (int j = 0; j < min(8, trainer_number[i]); ++j)
			cwin>>winner[i][j];
	}
	ifstream clate("late.txt", std::ifstream::in);
	set<string> late;
	string s;
	clate>>s;
	while (s != "end")
	{
		late.insert(s);
		clate>>s;
	}
	ifstream cpred("predicts.txt", std::ifstream::in);
	vector<map<string, int> > points(2);
	map<string, int> late_points;
	vector<set<string> > first(3);	
	for (int i = 0; i < 2; ++i)
	{
		cpred>>s;
		while (s != "end")
		{
			string name;
			while (s != "Турнир")
			{
				name += (s + " ");
				cpred>>s;
			}
			if (points[i].find(name) == points[i].end())
			{
				points[i][name] = 0;
				//if (trainer_number[i] < 8) points[i][name] = dif_to_points(0);
			}
			if (late_points.find(name) == late_points.end())
				late_points[name] = 0;
			while (s != "фазе\"")
				cpred>>s;
			int place;
			cpred>>place;
			cpred>>s;
			if (late.find(s) != late.end())
				++late_points[name];
			else
			{
				place -= late_points[name];
				if (place < 9)
				{
					for (int j = 0; j < min(8, trainer_number[i]); ++j)
					{
						if (s == winner[i][j])
						{
							points[i][name] += dif_to_points(abs(j - place + 1));
						}
					}
					if (place == 1 && winner[i][0] == s)
					{
						if (first[0].find(name) != first[0].end())
						{
							first[0].erase(first[0].find(name));
							first[2].insert(name);
						}
						else
							first[i].insert(name);
					}
				}
			}
			cpred>>s;
		}
		late_points.clear();
	}
	//freopen("output.txt", "w", stdout);
	for (int i = 0; i < 2; ++i)
	{
		for (map<string, int>::iterator it = points[i].begin(); it != points[i].end(); ++it)
			cout<<it->first<<" "<<it->second<<" "<<points_to_money(it->second, trainer_number[i])<<"\n";
		cout<<"\n";
	}

	set<string> legendary;
	set<string> pokemon;
	map<string, int> money;
	map<string, int> credits;
	for (int i = 0; i < 2; ++i)
	{
		for (map<string, int>::iterator it = points[i].begin(); it != points[i].end(); ++it)
		{
			if (legend_points(it->second, trainer_number[i]))
				legendary.insert(it->first);
			if (pokemon_points(it->second, trainer_number[i]))
				pokemon.insert(it->first);
			if (money.find(it->first) == money.end())
				money[it->first] = 0;
			money[it->first] += points_to_money(it->second, trainer_number[i]);
			if (credits.find(it->first) == credits.end())
				credits[it->first] = 0;
			credits[it->first] += points_to_credit(it->second, trainer_number[i]);
		}
	}
	map<int, vector<string> > money_list;
	empty.clear();
	for (map<string, int>::iterator it = money.begin(); it != money.end(); ++it)
	{
		if (money_list.find(it->second) == money_list.end())
			money_list[it->second] = empty;
		money_list[it->second].push_back(it->first);
	}
	map<int, vector<string> > credit_list;
	empty.clear();
	for (map<string, int>::iterator it = credits.begin(); it != credits.end(); ++it)
	{
		if (credit_list.find(it->second) == credit_list.end())
			credit_list[it->second] = empty;
		credit_list[it->second].push_back(it->first);
	}


	//freopen("prizes.txt", "w", stdout);
	if (! legendary.empty())
	{
		cout<<"Легенда + Лоск + титул \"Сорвавший джекпот\": ";
		for (set<string>::iterator it = legendary.begin(); it != legendary.end(); ++it)
			cout<<(*it)<<", ";
		cout<<"\n";
	}
	if (! pokemon.empty())
	{
		cout<<"Покемон + 1500 Поке-иен: ";
		for (set<string>::iterator it = pokemon.begin(); it != pokemon.end(); ++it)
			cout<<(*it)<<", ";
		cout<<"\n";
	}
	for (map<int, vector<string> >::iterator it = money_list.begin(); it != money_list.end(); ++it)
	{
		if (it->first > 0)
		{
			cout<<it->first<<" Поке-иен: ";
			for (int i = 0; i < it->second.size(); ++i)
				cout<<it->second[i]<<", ";
			cout<<"\n";
		}
	}
	for (map<int, vector<string> >::iterator it = credit_list.begin(); it != credit_list.end(); ++it)
	{
		if (it->first > 0)
		{
			cout<<it->first<<" Кредитов: ";
			for (int i = 0; i < it->second.size(); ++i)
				cout<<it->second[i]<<", ";
			cout<<"\n";
		}
	}
	if (! first[2].empty())
	{
		cout<<"3 Редких конфеты: ";
		for (set<string>::iterator it = first[2].begin(); it != first[2].end(); ++it)
			cout<<(*it)<<", ";
		cout<<"\n";
	}
	if (! first[0].empty())
	{
		cout<<"1 Редкая конфета (за победителя старшей лиги): ";
		for (set<string>::iterator it = first[0].begin(); it != first[0].end(); ++it)
			cout<<(*it)<<", ";
		cout<<"\n";
	}
	if (! first[1].empty())
	{
		cout<<"1 Редкая конфета (за победителя младшей лиги): ";
		for (set<string>::iterator it = first[1].begin(); it != first[1].end(); ++it)
			cout<<(*it)<<", ";
	}
	return 0;
}

File: guess.py
def dif_to_points(dif):
	return (dif * dif - 15 * dif + 58) // 2;

def legend_points(points, num):
	if num <= 8:
		return points >= 203
	if num in {9, 10}:
		return points >= 200
	if num in {11, 12}:
		return points >= 190
	if num >= 13:
		return points >= 180
	return False

def pokemon_points(points, num):
	if num <= 8:
		return points >= 190
	if num in {9, 10}:
		return points >= 180
	if num in {11, 12}:
		return points >= 165
	if num >= 13:
		return points >= 150
	return False

def points_to_money(prizes, points, num):
	if num <= 8:
		if points >= 203:
			return prizes[0]
		if points >= 190:
			return prizes[1]
		if points >= 175:
			return prizes[2]
		if points >= 165:
			return prizes[3]
		if points >= 155:
			return prizes[4]
	if num in {9, 10}:
		if points >= 200:
			return prizes[0]
		if points >= 180:
			return prizes[1]
		if points >= 160:
			return prizes[2]
		if points >= 150:
			return prizes[3]
		if points >= 140:
			return prizes[4]
	if num in {11, 12}:
		if points >= 190:
			return prizes[0]
		if points >= 165:
			return prizes[1]
		if points >= 140:
			return prizes[2]
		if points >= 130:
			return prizes[3]
		if points >= 120:
			return prizes[4]
	if num >= 13:
		if points >= 180:
			return prizes[0]
		if points >= 150:
			return prizes[1]
		if points >= 120:
			return prizes[2]
		if points >= 110:
			return prizes[3]
		if points >= 100:
			return prizes[4]
	return 0

def print_prize_list(prize, winners):
	if winners:
		print('%s: %s' % (prize, ', '.join(sorted(winners))))


tours = dict()
tours_list = []

guess_file = open("data/guess.txt")
guesses = []
for guess in guess_file:
	guesses.append(guess.rstrip().split('\t'))
	if len(guesses[-1]) > 1 and guesses[-1][1] not in tours:
		tours[guesses[-1][1]] = dict()
		tours_list.append(guesses[-1][1])
guess_file.close()

winner_file = open("data/winners.txt")
for name in tours_list:
	tours[name]['num'] = int(winner_file.readline().rstrip())
	tours[name]['winner'] = dict()
	for i in range(min(8, tours[name]['num'])):
		s = winner_file.readline()
		if '\t' in s:
			winner_name = s.split('\t')[1]
		else:
			winner_name = s
		tours[name]['winner'][winner_name.rstrip()] = i + 1
winner_file.close()

late_file = open("data/late.txt")
late = set()
for s in late_file:
	if len(s.rstrip()) > 0:
		late.add(s.rstrip())
late_file.close()

shift = dict()
points = dict()

guessed_winner = {'all': set()}
for tour in tours:
	guessed_winner[tour] = set()

for guess in guesses:
	if len(guess) >= 4:
		user = guess[0].rstrip()
		tour = guess[1]
		place = int(guess[2])
		trainer = guess[3].rstrip()
		if user in points:
			if tour not in points[user]:
				points[user][tour] = 0
				shift[user][tour] = 0
		else:
			points[user] = dict()
			points[user][tour] = 0
			shift[user] = dict()
			shift[user][tour] = 0

		if trainer in late:
			shift[user][tour] += 1
		elif trainer in tours[tour]['winner']:
			place -= shift[user][tour]
			if place <= 8:
				if trainer in {'',}:
					points[user][tour] += max(dif_to_points(abs(tours[tour]['winner'][trainer] - 6)), dif_to_points(abs(tours[tour]['winner'][trainer] - 7)))
				else:
					points[user][tour] += dif_to_points(abs(tours[tour]['winner'][trainer] - place))
			if tours[tour]['winner'][trainer] == 1 and place == 1:
				guessed_winner[tour].add(user)

print(tours_list, tours)

for tour in tours:
	print(tour)
	for user in points:
		if tour in points[user]:
			print(user, points[user][tour])

prize_file = open("data/prizes.txt")
prizes = {'candy': []}
for s in prize_file:
	num = int(s.split('. ')[0])
	s = ' ('.join(s.split('. ')[1].split(' (')[:-1])
	if num == 1:
		jackpot = s.split(' + ')
		prizes['legend'] = ' + '.join(jackpot[:-1])
		currency = ' '.join(jackpot[-1].split(' ')[1:])
		gain = int(jackpot[-1].split(' ')[0])
		prizes[currency] = [gain, 0, 0, 0, 0]
	elif num in {2, 3, 4, 5}:
		lot = s.split(' + ')
		for prize in lot:
			if num == 2 and prize == lot[0]:
				prizes['pokemon'] = prize
			else:
				currency = ' '.join(prize.split(' ')[1:])
				gain = int(prize.split(' ')[0])
				if currency not in prizes:
					prizes[currency] = [0, 0, 0, 0, 0]
				prizes[currency][num - 1] = gain

	elif num in {6, 7}:
		prizes['candy'].append(s.split(' (')[0])
prize_file.close()
if len(prizes['candy']) == 1:
	prizes['candy'].append(prizes['candy'][0])

legend = set()
pokemon = set()
money = dict()
for currency in prizes:
	if type(prizes[currency]) is list and currency != 'candy':
		money[currency] = dict()
for user in points:
	for tour in points[user]:
		if legend_points(points[user][tour], tours[tour]['num']):
			legend.add(user)
		elif pokemon_points(points[user][tour], tours[tour]['num']):
			pokemon.add(user)
		for currency in money:
			if user not in money[currency]:
				money[currency][user] = 0
			money[currency][user] += points_to_money(prizes[currency], points[user][tour], tours[tour]['num'])

money_list = dict()
for currency in money:
	money_list[currency] = dict()
	for user in money[currency]:
		if money[currency][user] > 0:
			if money[currency][user] in money_list[currency]:
				money_list[currency][money[currency][user]].append(user)
			else:
				money_list[currency][money[currency][user]] = [user]

print_prize_list(prizes['legend'], legend)
print_prize_list(prizes['pokemon'], pokemon - legend)

if len(tours) > 1:
	for user in points:
		guessed_winners = 0
		for tour in tours:
			if user in guessed_winner[tour]:
				guessed_winners += 1
		if guessed_winners == len(tours):
			guessed_winner['all'].add(user)
	print_prize_list(prizes['candy'][0] + ' (все лиги)', guessed_winner['all'])

for tour in tours:
	name = ''
	if len(tour.split(' - ')) > 1:
		name = '(%s)' % tour.split(' - ')[1] 
	print_prize_list(' '.join([prizes['candy'][1], name]), guessed_winner[tour] - guessed_winner['all'])

currency_list = sorted(list(money_list))
for currency in currency_list:
	money_list[currency] = sorted([(gain, money_list[currency][gain]) for gain in money_list[currency]], key=lambda x: -x[0])
	for i in range(len(money_list[currency])):
		print_prize_list('%d %s' % (money_list[currency][i][0], currency), money_list[currency][i][1])

